source/draw/gpu/android/AndroidContext.ooc

```diff
@@ -14,79 +14,82 @@
+				index = i
+				break
+			}
 		}
-		result
-	}
-	toRaster: func ~uv (gpuImage: GpuUv, async: Bool = false) -> RasterImage {
-		result: RasterImage
-		bytesPerPixel := 2
-		if (!this isAligned(gpuImage size width * bytesPerPixel))
-			result = gpuImage toRasterDefault()
-		else {
-			uvPacker := this createPacker(gpuImage size, bytesPerPixel)
-			this _packUv imageWidth = gpuImage size width
-			this _packUv channels = bytesPerPixel
-			uvPacker pack(gpuImage, this _packUv)
-			result = RasterUv new(uvPacker read(async), gpuImage size, 64)
+		if (index == -1) {
+			androidTexture := this createAndroidRgba(size)
+			result = OpenGLES3Bgra new(androidTexture, this)
+			result _recyclable = false
 		}
+		else
+			result = this _packers remove(index)
 		result
 	}
-	toRaster: override func (gpuImage: GpuImage, async: Bool = false) -> RasterImage {
-		result := match(gpuImage) {
-			case (i : GpuUv) => this toRaster(gpuImage as GpuUv, async)
-			case (i : GpuMonochrome) => this toRaster(gpuImage as GpuMonochrome, async)
-			case => super(gpuImage)
-		}
-		result
+	toBuffer: func (gpuImage: GpuImage, packMap: OpenGLES3MapPack, async: Bool) -> ByteBuffer {
+		packSize := IntSize2D new(gpuImage size width / (4 / gpuImage channels), gpuImage size height)
+		gpuRgba := this getPacker(packSize)
+		packMap imageWidth = gpuImage size width
+		gpuImage setMagFilter(false)
+		gpuImage setMinFilter(false)
+		gpuRgba canvas draw(gpuImage, packMap, Viewport new(gpuRgba size))
+		if(!async)
+			Fence new() sync() . clientWait() . free()
+		androidTexture := gpuRgba texture as AndroidTexture
+		sourcePointer := androidTexture lock(false)
+		buffer := ByteBuffer new(sourcePointer, androidTexture stride * androidTexture size height,
+			func (buffer: ByteBuffer) {
+				androidTexture unlock()
+				this recyclePacker(gpuRgba)
+			})
+		buffer
+	}
+	toRaster: func ~monochrome (gpuImage: GpuMonochrome, async: Bool) -> RasterImage {
+		buffer := this toBuffer(gpuImage, this _packMonochrome, async)
+		RasterMonochrome new(buffer, gpuImage size)
 	}
-	recycle: func ~GpuPacker (packer: GpuPacker) { this _packerBin add(packer) }
-	createPacker: func (size: IntSize2D, bytesPerPixel: UInt) -> GpuPacker {
-		result := this _packerBin find(size, bytesPerPixel)
-		if (result == null) {
-			version(debugGL) { Debug print("Could not find a recycled GpuPacker!")}
-			result = GpuPacker new(size, bytesPerPixel, this)
+	toRaster: func ~uv (gpuImage: GpuUv, async: Bool) -> RasterImage {
+		buffer := this toBuffer(gpuImage, this _packUv, async)
+		RasterUv new(buffer, gpuImage size)
+	}
+	toRaster: override func (gpuImage: GpuImage, async: Bool = false) -> RasterImage {
+		result: RasterImage = null
+		if (!this isAligned(gpuImage channels * gpuImage size width))
+			result = super(gpuImage, async)
+		else {
+			result = match(gpuImage) {
+				case (i : GpuUv) => this toRaster(gpuImage as GpuUv, async)
+				case (i : GpuMonochrome) => this toRaster(gpuImage as GpuMonochrome, async)
```

Comment:
Pretty sure this will give a warning for `i` being unused. Try
```ooc
	result = match(gpuImage) {
	case (gpuImage: GpuUv) => this toRaster(gpuImage as GpuUv, async)
	case (gpuImage: GpuMonochrome) => this toRaster(gpuImage as GpuMonochrome, async)
```